Remotion/Data/DomainObjects/Mapping/ClassDefinition.cs

```diff
@@ -120,61 +90,39 @@ public bool IsSameOrBaseClassOf (ClassDefinition classDefinition)
       return false;
     }
 
-    public ClassDefinition[] GetAllDerivedClasses ()
+    public override TypeDefinition[] GetTypeHierarchy ()
     {
       var allDerivedClasses = new List<ClassDefinition>();
+      allDerivedClasses.Add(this);
       FillAllDerivedClasses(allDerivedClasses);
-      return allDerivedClasses.ToArray();
-    }
-
-    public ClassDefinition GetInheritanceRootClass ()
-    {
-      if (BaseClass != null)
-        return BaseClass.GetInheritanceRootClass();
-
-      return this;
-    }
 
-    [Obsolete("Contains (...) method was ambiguous between GetPropertyDefinitions() and MyPropertyDefinitions. Use GetPropertyDefinitions().Contains (...) instead. (Version: 3.0.0)", true)]
-    public bool Contains (PropertyDefinition propertyDefinition)
-    {
-      throw new NotSupportedException("Use GetPropertyDefinitions().Contains (...) instead. (Version: 3.0.0)");
-    }
+      var allDerivedTypes = new TypeDefinition[allDerivedClasses.Count];
+      for (var i = 0; i < allDerivedTypes.Length; i++)
+        allDerivedTypes[i] = allDerivedClasses[i];
```

Comment:
Is this an Array.Copy manually implemented? -> switch to Array.Copy ?

Reply:
That won't work because one of them is a `List<T>`. This could be replaced with a LINQ expression but that feels so unnecessary.

Reply:
`allDerivedTypes = allDerivedClasses.ToArray()` then?

Reply:
It would need to be `allDerivedClasses.Cast<TypeDefinition>().ToArray()` as the type of the list is `ClassDefinition`. And I did not like that which is why I manually wrote the copy loop.

Reply:
hmm... I see.
How about we just change the returntype from arry to IReadOnlyCollection (or IReadOnlyList)? That way, we don't need to copy and don't need to type-change?

Reply:
I don't think that returning the `List<T>` is better than reallocating the exact amount of space needed and copying. So I would either change it to the LINQ version or keep the copy as is.

Reply:
Am confused. What's the problem with returning an IReadOnlyList<T> (or IReadOnlyCollection<T> if no indexer is needed). We already build a new list during the execution and then it's a readonly api that gets returned. This would actually be a perfect fit from my POV. What am I missing?

Reply:
If I understood correctly, you are proposing we change the return type to `IReadOnlyCollection<TypeDefinition>` and simply return the `List<ClassDefinition>`. I dislike returning `List<T>`s as they might occupy a lot more space than necessary. As such, I feel like creating a new array is always preferred.

I don't know what problem we are actually trying to solve with this thread. Originally the question was about `Array.Copy` not being used, which was not used as it could not be used in that scenario. I assume you want to get rid of the `for` loop? If so why?

Reply:
Yeah, the for-loop triggered me because it's a lot of code that should have a benefit. (I'm perfectly happy to optimize code in tight loop scnearios but here, it just feels not worth it to write code that uses indexers and counters instead of something declarative)

The change of the return value is in the end is simply the final evolution that resulted from learning the less invloved approaches don't work. However, now that I see the final picture, changing the return value and returning a List<T> as an IReadOnlyCollection is actually the cleanest solution from my perspective because it removes complexity from the code. 

Regarding the memory requirement of List<T>: yes, the list grows with a power of two (thanks for reminding me, I do tend to ignore the excess space). In our scenarios: Typically, there's perhaps 500 class types in the entire domain, so we're not in a problematic territory by far. The list must be allocated anyway, which means we only save memory during the time the result object (as an array instead of a list) is being used at the call site. From a pratical POV, the result will be iterated/processed and then discared. Which means, we most likely generate less memory pressure by not copying the list and we definitely safe some cpu cycles by not copying the array. We could discuss a different approach once we decide to lazy cache the result (which may be a good idea eventually).

Reply:
Historically speaking: the array-result was there because way back when, that was the only way to return a strong typed collection without implementing a collection type youself. So, the switch to IReadOnlyCollection is actually api cleanup to get rid of the nasty arrays.